virtual_ecosystem/core/core_components.py

```diff
@@ -188,21 +230,33 @@ class LayerStructure:
 
     canopy_layers: int = field(init=False)
     """The maximum number of canopy layers."""
-    soil_layers: list[float] = field(init=False)
+    soil_layers: NDArray[np.float32] = field(init=False)
     """A list of the depths of soil layer boundaries."""
+    soil_layer_thickness: NDArray[np.float32] = field(init=False)
+    """A list of the thicknesses of soil layers."""
     above_canopy_height_offset: float = field(init=False)
     """The height above the canopy of the provided reference climate variables."""
     surface_layer_height: float = field(init=False)
     """The height above ground used to represent surface conditions."""
-    subcanopy_layer_height: float = field(init=False)
-    """The height above ground used to represent subcanopy conditions."""
-    layer_roles: list[str] = field(init=False)
-    """An tuple of the roles of the vertical layers within the model from top to
-    bottom."""
     n_layers: int = field(init=False)
     """The total number of vertical layers in the model."""
+    layer_indices: NDArray[np.int_] = field(init=False)
+    """The numeric indices of the vertical layers in the model."""
+    layer_roles: NDArray[np.str_] = field(init=False)
+    """An array of the roles of the vertical layers within the model from top to
+    bottom."""
+    role_indices_bool: dict[str, NDArray[np.bool_]] = field(init=False)
+    """A dictionary providing boolean indices of each role within the vertical layer
+    structure."""
+    role_indices: dict[str, NDArray[np.int_]] = field(init=False)
+    """A dictionary providing integer indices of each roles within the vertical layer
+    structure."""
     config: InitVar[Config]
     """A validated model configuration."""
+    n_cells: int
+    """The number of grid cells in the simulation."""
+    max_depth_of_microbial_activity: float
+    """The maximum depth of microbial activity (m)."""
 
     def __post_init__(self, config: Config) -> None:
```

Comment:
This post init method is pretty long and doing a lot of stuff. I suggest breaking it down in smaller chunks for:
- validation
- setting roles and indices
- set microbial information

Reply:
That's true, but there are a lot of attributes, which should be declared and documented in `__init__` rather than other methods. I could simply declare and type them all and then populate them in sub methods, but most of them are populated in one liners. It makes clear sense to break out initialising the indices though.